File: app/src/main/java/juniojsv/minimum/SearchApps.java
package juniojsv.minimum;

import android.content.Intent;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.graphics.drawable.Drawable;
import android.os.AsyncTask;

import java.util.ArrayList;
import java.util.List;

import juniojsv.minimum.Utilities.SortListOfApps;

public class SearchApps extends AsyncTask<Void, Void, List<App>> {
    private PackageManager packageManager;
    private MinimumInterface minimumInterface;

    SearchApps(PackageManager packageManager, MinimumInterface minimumInterface) {
        this.packageManager = packageManager;
        this.minimumInterface = minimumInterface;
    }

    @Override
    protected void onPreExecute() {
        super.onPreExecute();
        minimumInterface.onSearchAppsStarting();
    }

    @Override
    protected List<App> doInBackground(Void... voids) {

        List<ApplicationInfo> installedApps = packageManager.getInstalledApplications(0);
        List<App> newAppsList = new ArrayList<>();

        for (ApplicationInfo app : installedApps) {
            String packageLabel = app.loadLabel(packageManager).toString();
            Drawable icon = app.loadIcon(packageManager);
            Intent intent = packageManager.getLaunchIntentForPackage(app.packageName);
            String packageName = app.packageName;

            if (intent != null && !app.packageName.equals(BuildConfig.APPLICATION_ID)) {

                intent.setAction(Intent.ACTION_MAIN);
                intent.addCategory(Intent.CATEGORY_LAUNCHER);

                newAppsList.add(new App(packageLabel, icon, intent, packageName));
            }
        }

        new SortListOfApps(newAppsList);

        return newAppsList;
    }

    @Override
    protected void onPostExecute(List<App> newAppsList) {
        super.onPostExecute(newAppsList);
        minimumInterface.onSearchAppsFinished(newAppsList);
    }
}

File: app/src/main/java/juniojsv/minimum/SearchApps.kt
package juniojsv.minimum

import android.content.Intent
import android.graphics.drawable.Drawable
import android.os.AsyncTask
import juniojsv.minimum.utilities.SortListOfApps
import java.util.*

class SearchApps internal constructor(private val minimum: MinimumActivity) : AsyncTask<Void, Void, MutableList<App>>() {
    private val packageManager = minimum.packageManager

    override fun onPreExecute() {
        super.onPreExecute()
        minimum.onSearchAppsStarting()
    }

    override fun doInBackground(vararg voids: Void): MutableList<App>? {

        return ArrayList<App>().apply {
            packageManager.getInstalledApplications(0).forEach {

                val intent: Intent? = packageManager.getLaunchIntentForPackage(it.packageName)

                if (intent != null && it.packageName != BuildConfig.APPLICATION_ID) {
                    intent.apply {
                        action = Intent.ACTION_MAIN
                        addCategory(Intent.CATEGORY_LAUNCHER)
                    }

                    val packageLabel: String = it.loadLabel(packageManager).toString()
                    val icon: Drawable = it.loadIcon(packageManager)
                    val packageName: String = it.packageName

                    add(App(packageLabel, icon, intent, packageName))
                }
            }
            SortListOfApps(this)
        }
    }

    override fun onPostExecute(newAppsList: MutableList<App>) {
        super.onPostExecute(newAppsList)
        minimum.onSearchAppsFinished(newAppsList)
    }
}